docs/source/custom_models.mdx

```diff
@@ -0,0 +1,153 @@
+of the repository with no abstraction, so you can easily copy a modeling file and tweak it to your needs.
+
+Once you are happy with those tweaks and trained a model you want to share with the community, there are simple steps
+to push on the Model Hub not only the weights of your model, but also the code it relies on, so that anyone in the
+community can use it, even if it's not present in the 🤗 Transformers library.
+
+This also applies to configurations and tokenizers (support for feature extractors and processors is coming soon).
+
+## Sending the code to the Hub
+
+First, make sure your model is fully defined in a `.py` file. It can rely on relative imports to some other files as
+long as all the files are in the same directory (we don't support submodules for this feature yet). For instance,
+let's say you have a `modeling.py` file and a `configuration.py` file in a folder of the current working directory
+named `awesome_model`, and that the modeling file defines an `AwesomeModel`, the configuration file a `AwesomeConfig`.
+
+```
+.
+└── awesome_model
+    ├── __init__.py
+    ├── configuration.py
+    └── modeling.py
+```
+
+The `__init__.py` can be empty, it's just there so that Python detects `awesome_model` can be use as a module.
+Here is an example of what the configuration file could look like:
+
+```py
+from transformers import PretrainedConfig
+
+
+class AwesomeConfig(PretrainedConfig):
+    model_type = "awesome"
+
+    def __init__(self, attribute=1, **kwargs):
+        self.attribute = attribute
+        super().__init__(**kwargs)
+```
+
+and the modeling file could have content like this:
+
+```py
+import torch
+
+from transformers import PreTrainedModel
+
+from .configuration import AwesomeConfig
+
+
+class AwesomeModel(PreTrainedModel):
+    config_class = AwesomeConfig
+    base_model_prefix = "base"
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.linear = torch.nn.Linear(config.hidden_size, config.hidden_size)
+
+    def forward(self, x):
+        return self.linear(x)
+```
+
+`AwesomeModel` should subclass [`PreTrainedModel`] and `AwesomeConfig` should subclass [`PretrainedConfig`]. The
+easiest way to achieve this is to copy the modeling and configuration files of the model closest to the one you're
+coding, and then tweaking them.
+
+Note that you can re-use (or subclass) an existing configuration/model.
+
+To share your model with the community, follow those steps: first import the custom objects.
+
+```py
+from awesome_model.configuration import AwesomeConfig
+from awesome_model.modeling import AwesomeModel
+```
+
+Then you have to tell the library you want to copy the code files of those objects when using the `save_pretrained`
+method and properly register them with a given Auto class (especially for models), just run:
+
+```py
+AwesomeConfig.register_for_auto_class()
+AwesomeModel.register_for_auto_class("AutoModel")
```

Comment:
Is it possible to validate the auto class str here, in case there is a typo? Right now it's possible to do
`AwesomeConfig.register_for_auto_class("auto")` and it won't fail. Or does it fail down the line?

Reply:
It will fail when the user tries to load the model. Can add a check!